### ISSUE TITLE
Create Integration Tests "infrastructure"

### ISSUE THREAD

**Issue by kptdobe** [adobe/helix-pages#8]
An important piece of the project is the customisation of the VCL code in order to extract the owner / repo / branch from the subdomain. Unfortunately, VCL code cannot be unit tested and requires some tricky regex.

Goal here is to write IT that validates against the Fastly environment that the different url patterns are recognised as expected. It is easy to construct urls that must return something assuming the underlying Git projects do not disappear:

https://hlxtest-tripodsan.project-helix.page/
https://hlxtest--tripodsan.project-helix.page/
https://hello-helix--stefan-guggisberg.project-helix.page/README.html
https://helix-test-davidnuescheler.project-helix.page/5-bsl.html
https://helix-test--davidnuescheler.project-helix.page/team.html
https://to-helix-pages--helix-test--davidnuescheler.project-helix.page/

We should probably create a stable Git repo for each of the test case (owner, repo, branch, with and without a dash) that we can test against.

**Comment by kptdobe** [comment]
In order to run those IT tests, we would need:

- [ ] a `helix-pages-testscontent` git repo
  - The repo could contain different flavours of the content on which predictable tests could be run against
    - [ ] a simple `index.md`
    - [ ] multiple branches representing different content variations like a different `index.md` on a branch `v2` 
    - [ ] multiple branches containing code / static overrides like `style.css` override
- [ ] a `testsproject-helix.page` domain with the corresponding Fastly account.

Running the IT test would mean:
- deploy / publish to the `helix-pages-test.page` domain/service
- a bunch of requests like:
  - https://www.testsproject-helix.page (at least should respond some html)
  - https://helix-pages-testscontent-adobe.testsproject-helix.page/ (predicate html)
  - https://v2--helix-pages-testscontent--adobe.testsproject-helix.page/ (predicate html different from previous one)
  - https://cssoverride--helix-pages-testscontent--adobe.testsproject-helix.page/style.css (predicable css) which is also different from https://helix-pages-testscontent-adobe.testsproject-helix.page/style.css (helix-pages project css)

Like this will also be able to test various overrides capabilities introduced by upcoming https://github.com/adobe/helix-publish/issues/96

@tripodsan @trieloff WDYT ? 
If you agree, we'll need a new domain name :)

**Comment by trieloff** [comment]
Looks good. In order to reduce parallelism, we could potentially use [this API](https://circleci.com/docs/api/#recent-builds-for-a-single-project) to make sure there are no other jobs running (and wait five minutes otherwise)

We could even increase the level of parallelism by acquiring `helix-[1-n].dev` (1, 2, and 3 are still available) and then use the modulo of the build number to pick the domain name.

**Comment by bdelacretaz** [comment]
We discussed this today at the SXB hackathon, here's our suggested plan for testing changes to Helix Pages code or any of its dependencies:
 * Deploy the current Helix Pages code to a temporary domain
* Look for GitHub repositories which have a `helix-pages-it` topic
* For each such repository, run the integration tests against that temporary domain, using a standardized command, suggest `npm run helix-pages-it`
* Report on the aggregated test results

This allows us to integrate any Helix Pages repository in those integration tests, including ones from customers if appropriate.

The whole process can become part of the `helix-continuous` smoke tests.

**Comment by bdelacretaz** [comment]
I've done a small experiment for running the tests of many repositories in a single CircleCI build.

See [circleci-multiple-tests/17](https://circleci.com/gh/bdelacretaz/circleci-multiple-tests/17) for an example that runs `npm test` on four Helix repositories, including one that fails. Source code at [bdelacretaz/circleci-multiple-tests](https://github.com/bdelacretaz/circleci-multiple-tests)

Each repo is in a single CircleCI step, so results are well organized, and I think multiple repo tests can run in parallel if the CircleCI account allows this.

A failing repo like `helix-example-basic` does not stop the tests.

I think this would be an option for running tests on a set of tagged repos that are discovered in a pre-test phase, though in my case the [.circleci/config.yml](https://github.com/bdelacretaz/circleci-multiple-tests/blob/master/.circleci/config.yml) is static, it contains definitions like:

      - run:
          name: test helix-static
          command: npm run test-repo git@github.com:adobe/helix-static.git
          when: always
      - run:
          name: test helix-example-basic
          command: npm run test-repo git@github.com:adobe/helix-example-basic.git
          when: always
     ....

...but those could be generated.

**Comment by kptdobe** [comment]
Even better: start a workflow per test (not a job per test). Like this you get a nice diagram like https://circleci.com/workflow-run/f7e9848d-fc29-4cd5-83ab-da6c4189a699 (vs one box diagram like your https://circleci.com/workflow-run/cbaa8a06-799b-4860-ba35-9047183d6a4a.

**Comment by kptdobe** [comment]
Also, with you model, you do not benefit from parallelism.

The conclusion of my researches is exactly around the `those could be generated`: there is no way to make the workflows dynamic. You need to have the full list of things you want to run in the config if you want a readable diagram. Btw, a readable diagram is important because it is easy to know where it failed. With a generated approach, today, the only solution is a bash script which will call all the `npm run test`: the statuses will be in the middle of that bash script execution.
This is something we had for the smoke tests and with 2 jobs, people were already lost (i.e. could not find why it fails). Imagine with more...

**Comment by kptdobe** [comment]
I am working on a proposal for `helix-pages`, should be ready tomorrow or Wednesday.

**Comment by kptdobe** [comment]
Btw, it seems that buidkite can do what we are looking for: https://buildkite.com/docs/pipelines/defining-steps#dynamic-pipelines

I still want to test it.

**Comment by bdelacretaz** [comment]
>  ...there is no nice way to make the workflows dynamic...

And generating a set of workflow definitions like you have in the [helix-continuous/config.yml](https://github.com/adobe/helix-continuous/blob/master/.circleci/config.yml) looks like a workable solution to me. The test commands and other variations for each repository can be generated from a `.helix` file in the repository, using sensible defaults.

**Comment by kptdobe** [comment]
Each time I think I found the way to go, I find a new hurdle.
I think this is a good exercise for me to summarise here what my current plan is:

In Helix Pages:
1. have a "Execute a project Helix Pages IT specific tests" generic job on which you can configure the repo name. This job will run the tests store inside the project and give it the test domain to run against (like hlx-1.page, hlx-2.page...)
2. **manually** maintain a list of projects we want to run the tests against (like helix-pages itself, helix-example-basic, helix-example-advanced, lr.landingpage, ferrum). Manually means in the Helix Pages `.circleci/config.yaml`, have a list of 5 jobs describe in 5 with the appropriate repo names. While see later how we can do the dynamic part. I start doubting about a full dynamic version of it because: 
- performances (the idea is to run the smoke tests on each commit in our helix repos, we should keep that under control)
- a failing test during the dev process is time lost, the repos being tested must not be any random repo, at least for now. The decision to had to `.circleci/config.yaml` is a good gatekeeper
In Helix repos (cli, pipeline...) via the smoke-test orb:
1. extend the current smoke tests and add a new box: helix-pages
2. the job uses the "patched" helix cli (the one with the current branch code being developed) to deploy / publish Helix Pages master on a test domain (hlx-1.page, hlx-2.page... rotation)
3. the job triggers the Helix Pages smoke tests describe above and waits for execution.

This looks really complicated :)

**Comment by kptdobe** [comment]
After reviewing my writing, I found a new issue: the test domain rotation would need to happen from 2 ci: helix-pages and helix-continuous. This will not work because the risk of override is too big.

**Comment by kptdobe** [comment]
Here is an example of what I imagined: https://circleci.com/workflow-run/9f9ab284-ed2c-405b-ab3f-6d576423d94f
Corresponding config: https://github.com/adobe/helix-pages/blob/smoke-tests/.circleci/config.yml

CircleCI parametrized job (poor doc) allows us to create a generic job and to have multiple instances with different parameters (one instance per smoke test we want to run).

**Comment by kptdobe** [comment]
And the smoke tests run in parallel.

**Comment by kptdobe** [comment]
Created PR https://github.com/adobe/helix-pages/pull/99 with the initial proposal - everything described in PR.

**Comment by kptdobe** [comment]
Merged and released.

Follow up tasks:
- #102 
- #103 
- #104
- https://github.com/adobe/helix-continuous/issues/43
